File: 027_NumberOfRecentCalls.cpp
// Topic: Queue / Class
// Number of Recent Calls: Count pings that occurred in the last 3000 milliseconds.
// Time Complexity: O(1) amortized per ping
// Space Complexity: O(n) for stored pings

#include <bits/stdc++.h>
using namespace std;

class RecentCounter
{
   public:
    deque<int> q;
    RecentCounter() {}
    int ping(int t)
    {
        q.push_back(t);
        while (!q.empty() && q.front() < t - 3000) q.pop_front();
        return q.size();
    }
};

int main()
{
    RecentCounter rc;
    cout << rc.ping(1) << '\n';     // 1
    cout << rc.ping(100) << '\n';   // 2
    cout << rc.ping(3001) << '\n';  // 3
    cout << rc.ping(3002) << '\n';  // 3
    cout << rc.ping(7000) << '\n';  // 1
    return 0;
}

File: 027_NumberOfRecentCalls.py
"""
Topic: Queue / Sliding Window
Number of Recent Calls: implement class with queue storing timestamps
Time Complexity: O(1) amortized per call
Space Complexity: O(n) where n is number of recent calls
"""

from collections import deque

class RecentCounter:
    def __init__(self):
        self.q = deque()
    def ping(self, t: int) -> int:
        self.q.append(t)
        while self.q and self.q[0] < t - 3000:
            self.q.popleft()
        return len(self.q)


if __name__ == '__main__':
    rc = RecentCounter()
    print(rc.ping(1))    # 1
    print(rc.ping(100))  # 2
    print(rc.ping(3001)) # 3
